File: 백준/Gold/1202. 보석 도둑/보석 도둑.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Gem implements Comparable<Gem>{
	int w, v;
	
	public Gem(int w, int v) {
		this.w = w;
		this.v = v;
	}

	@Override
	public int compareTo(Gem o) { // 보석 무게 기준 오름차
		return Integer.compare(this.w, o.w);
	}
}

public class Main{
	
	public static int N, K;
	public static Gem[] gem;
	public static int[] bag;
	public static long ans;
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		
		gem = new Gem[N];
		bag = new int[K];
		
		int w, v;
		for(int i=0; i<N; i++) {
			st = new StringTokenizer(br.readLine());
			
			w = Integer.parseInt(st.nextToken());
			v = Integer.parseInt(st.nextToken());
			
			gem[i] = new Gem(w, v);
		}
		
		for(int i=0; i<K; i++) {
			bag[i] = Integer.parseInt(br.readLine());
		}
		
		// ===== solve =====
		
		ans = 0;
		
		Arrays.sort(gem);
		Arrays.sort(bag);
		
		PriorityQueue<Integer> pq = new PriorityQueue<>(); // 최소 
		
		int idx = 0;
		for(int i=0; i<K; i++) {
			while(idx < N && gem[idx].w <= bag[i]) pq.offer(-gem[idx++].v);
			if(!pq.isEmpty()) ans -= pq.poll();
		}
		
		System.out.println(ans);
	}
}

File: 백준/Gold/1202. 보석 도둑/보석 도둑.cc
#include <iostream>
#include <algorithm>
#include <queue>

#define MAX_N 300000
#define MAX_K 300000

using namespace std;

int N, K;
pair<int, int> gem[MAX_N];
int bag[MAX_K];
bool isTaken[MAX_N];
long long ans;

bool compare(pair<int, int> &a, pair<int, int> &b){
    return a.first < b.first;
}

void solve(){
    sort(bag, bag + K); // 오름차순
    sort(gem, gem + N, compare); // 보석 무게 기준 내림차순
    
    priority_queue<int> pq;
    
    int idx = 0;
    for(int i=0; i<K; i++){
        while(idx < N && gem[idx].first <= bag[i]) pq.push(gem[idx++].second);
        if(pq.empty()) continue;
        ans += pq.top();
        pq.pop();
    }
    
    return;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    cin >> N >> K;
    
    for(int i=0; i<N; i++) cin >> gem[i].first >> gem[i].second;
    for(int i=0; i<K; i++) cin >> bag[i];
    
    solve();
    
    cout << ans;
    
    return 0;
}
